hooks/useCachedImageURI.ts

```diff
@@ -0,0 +1,69 @@
+import {QueryClient, useQuery} from '@tanstack/react-query';
+import * as FileSystem from 'expo-file-system';
+import md5 from 'md5';
+import Log from 'network/log';
+
+export const useCachedImageURI = (uri: string) => {
+  return useQuery<string, Error>({
+    queryKey: queryKey(uri),
+    queryFn: async () => fetchCachedImageURI(uri),
+    cacheTime: 24 * 60 * 60 * 1000, // hold on to this cached data for a day (in milliseconds)
+    staleTime: 24 * 60 * 60 * 1000, // do not refresh this data
+  });
+};
+
+function queryKey(uri: string) {
+  return ['image', {uri: uri}];
+}
+
+const fetchCachedImageURI = async (uri: string) => {
+  if (!uri.startsWith('http')) {
+    // this is already a local file, no need to download
+    return uri;
+  }
+  const destination = FileSystem.cacheDirectory + '/' + md5(uri);
```

Comment:
suggestion: add a date component to the path (`2022-02-21`) and then at startup you can prune stale directories efficiently, in case things get missed by the cleanup routine

Reply:
also, what happens to `destination` here? how does a later request for `uri` get turned into a filesystem read of `destination`? I think there's some magic I'm missing

Reply:
So in general, I do not want to be in the business of cache management. What we do is that we store the _pointer_ in the `react-query` store, e.g.:

```
['image', {uri: 'https://...'}] => 'file://'
```

This is stored in memory and dehydrated into async storage. The normal cache TTL rules apply for the lifecycle of this _pointer_ -- we operate via side-effects in the hook and clean-up function: the actual query function side-effect places the file on disk before returning `destination` (which is the same as `result.uri`), and the cleanup function side-effect removes the file when the pointer is removed. Does that make sense?

Reply:
Oh, one other super critical bit - we are storing filesystem things in the cache directory, which is a location that the host OS can clean up whenever it wants, if the phone is low on storage, so we can't negatively impact the system here.

To that end, removing a file from disk (even based on dates) if `react-query` does not think it is removed, would lead to us returning the pointer to an empty file.

However, now that I think about it, we could have the same issue with the system removing parts of our cache under filesystem pressure, so I think a complete solution here needs to reconcile the filesystem state with the `react-query` state. Cache coherence is hard.

Reply:
It doesn't look like it's possible to watch the filesystem, so the best we can do is a periodic poll to reconcile the two caches.

Reply:
No, it's not guaranteed, but because we can't harm the system, I simply did not care -  more logic for no real benefit there :)

I've written a simple reconciler that we can run in the background anyway for the case where the system removed something, so we should be good here.

> If you're only deleting old files at startup (which is what I suggested) then I don't see how it could impact react-query.

`react-query` stores the cache state in async (persistent) storage, so when the app starts after a) being killed or b) crashing, it re-hydrates the cache, and any file links that were persisted but had their files removed would be invalid.